be/src/exec/spill/spill_components.cpp

```diff
@@ -0,0 +1,352 @@
+    if (state->is_cancelled()) {
+        return Status::OK();
+    }
+
+    const auto& serde = _spiller->serde();
+    spill::AcquireBlockOptions opts;
+    opts.query_id = state->query_id();
+    opts.plan_node_id = options().plan_node_id;
+    opts.name = options().name;
+    ASSIGN_OR_RETURN(auto block, _spiller->block_manager()->acquire_block(opts));
+
+    // TODO: reuse io context
+    SerdeContext spill_ctx;
+    // DCHECK(writable != nullptr);
+    {
+        SCOPED_TIMER(_spiller->metrics().write_io_timer);
+        // flush all pending result to spilled files
+        size_t num_rows_flushed = 0;
+        RETURN_IF_ERROR(mem_table->flush([&](const auto& chunk) {
+            num_rows_flushed += chunk->num_rows();
+            RETURN_IF_ERROR(serde->serialize(spill_ctx, chunk, block));
+            return Status::OK();
+        }));
+        TRACE_SPILL_LOG << "spill flush rows:" << num_rows_flushed << ",spiller:" << this;
+    }
+
+    // be careful close method return a not ok status
+    // then release the pending memory
+    // flush
+    RETURN_IF_ERROR(block->flush());
+    _spiller->block_manager()->release_block(block);
+
+    {
+        std::lock_guard<std::mutex> l(_mutex);
+        _block_group.append(std::move(block));
+    }
+
+    return Status::OK();
+}
+
+Status RawSpillerWriter::acquire_stream(const SpillPartitionInfo* partition,
+                                        std::shared_ptr<SpillInputStream>* stream) {
+    return acquire_stream(stream);
+}
+
+Status RawSpillerWriter::acquire_stream(std::shared_ptr<SpillInputStream>* stream) {
+    std::shared_ptr<SpillInputStream> input_stream;
+    const auto& serde = _spiller->serde();
+    const auto& opts = options();
+
+    if (opts.is_unordered) {
+        ASSIGN_OR_RETURN(input_stream, _block_group.as_unordered_stream(serde));
+    } else {
+        ASSIGN_OR_RETURN(input_stream,
+                         _block_group.as_ordered_stream(_runtime_state, serde, opts.sort_exprs, opts.sort_desc));
+    }
+
+    *stream = input_stream;
+
+    if (_mem_table != nullptr && !_mem_table->is_empty()) {
+        DCHECK(opts.is_unordered);
+        ASSIGN_OR_RETURN(auto mem_table_stream, _mem_table->as_input_stream(opts.read_shared));
+        *stream = SpillInputStream::union_all(mem_table_stream, *stream);
+    }
+
+    return Status::OK();
+}
+
+PartitionedSpillerWriter::PartitionedSpillerWriter(Spiller* spiller, RuntimeState* state)
+        : SpillerWriter(spiller, state), _mem_tracker(std::make_unique<MemTracker>(-1)) {}
+
+Status PartitionedSpillerWriter::prepare(RuntimeState* state) {
+    DCHECK_GT(options().init_partition_nums, 0);
+    _partition_set.resize(1024);
+    RETURN_IF_ERROR(_init_with_partition_nums(state, options().init_partition_nums));
+    for (auto [_, partition] : _id_to_partitions) {
+        RETURN_IF_ERROR(partition->spill_writer->prepare(state));
+        partition->spill_writer->acquire_mem_table();
```

Comment:
what will happen if we can't acquire_mem_table? _acquire_mem_table_from_pool may return nullptr

Reply:
see the implements in is_full(). we won't flush multi-mem table. so acquire_mem_table won't return nullptr